src/katsdpmodels/primary_beam.py

```diff
@@ -234,7 +323,437 @@ def sample_grid(self, l: ArrayLike, m: ArrayLike, frequency: u.Quantity,   # noq
+            conj = np.conj(values)
+            # Take Kronecker product. Unfortunately np.kron doesn't allow
+            # operating over subsets of dimensions.
+            M = np.block([[values[..., 0:1, 0:1] * conj, values[..., 0:1, 1:2] * conj],
+                          [values[..., 1:2, 0:1] * conj, values[..., 1:2, 1:2] * conj]])
+            return np.matmul(_XY_TO_IQUV @ M, _IQUV_TO_XY, out=out)
+        elif output_type == OutputType.UNPOLARIZED_POWER:
+            # Compute sum of squared magnitudes across the 4 Jones terms.
+            # Viewing at float32 simplifies summing squared magnitudes.
+            assert values.dtype == np.dtype(np.complex64)
+            ret = np.sum(np.square(values.view(np.float32)), axis=(-2, -1), out=out)
+            ret *= 0.5
+            return ret
+        else:
+            raise ValueError(f'Unrecognised output_type {output_type}')
+
+    @staticmethod
+    @numba.njit
+    def _sample_impl(aperture: np.ndarray, xf: np.ndarray, yf: np.ndarray,
+                     l: np.ndarray, m: np.ndarray, out: np.ndarray) -> None:
+        """Numba implementation details of :meth:`_sample_altaz_jones`.
+
+        Parameters
+        ----------
+        aperture
+            Aperture-plane samples, already interpolated onto the desired frequencies
+        xf, yf
+            x and y in wavelength, with the frequency axes first.
+        l, m
+            1D l and m coordinates (already broadcast with each other)
+        out
+            Output array
+        """
+        for freq_idx in np.ndindex(xf.shape[:-1]):
+            # Frequency-specific 1D arrays
+            x = xf[freq_idx]
+            y = yf[freq_idx]
+            ap = aperture[freq_idx]
+            coeff1 = _expjm2pi(np.outer(l, x))
+            coeff2 = _expjm2pi(np.outer(m, y))
+            for i in range(2):
+                for j in range(2):
+                    tmp = coeff2 @ ap[i, j]
+                    out_chunk = out[freq_idx + (...,) + (i, j)]
+                    for lm_idx in np.ndindex(l.shape):
+                        out_chunk[lm_idx] = np.dot(tmp[lm_idx], coeff1[lm_idx])
+
+    def _sample_altaz_jones(
+            self, l: np.ndarray, m: np.ndarray, frequency: u.Quantity,
+            frame: Union[AltAzFrame, RADecFrame], output_type: OutputType, *,
+            out: Optional[np.ndarray] = None) -> np.ndarray:
+        """Partial implementation of :meth:`sample`.
+
+        It takes `l` and `m` in AltAz frame, and produces Jones matrices in
+        either :data:`OutputType.JONES_HV` or :data:`OutputType.JONES_XY`.
+        The provided `frame` and `output_type` are used only for polarization
+        rotation.
+
+        l and m must already be broadcast to the same shape, and must be float32.
+        """
+        assert l.shape == m.shape
+        assert l.dtype == np.float32
+        assert m.dtype == np.float32
+
+        out_shape = frequency.shape + l.shape + (2, 2)
```

Comment:
expected this line to be `out_shape = frequency.shape + l.shape  + m.shape + (2, 2)`

Reply:
This is `sample`, not `sample_grid`, so we just have a list of (l, m) coordinates, not a Cartesian product.